### PR TITLE
Allow to create an Account from a private key

### PR DESCRIPTION
**Pull request by bcoste**
Allow to create an Account from a private key because it's not always possible to have the seed available and it's safer/faster to derive from the account hardened key.

### CODE REVIEW

**Comment by bcoste** [comment]
For the failing tests :
- testGetBaseAddressFromMnemonicByNetworkInfoMainnet 
- testGetBaseAddressFromMnemonicByNetworkInfoTestnet
- testGetBaseAddressFromMnemonicIndex_1
- testGetBaseAddressFromMnemonicIndex_2

This is due to my changes in the baseAddress, I force to calculate it from the external address of index 0.

From my understanding, base address represents the base index 0 of address_index in a BIP44 (and also CIP 1852) path :
m / purpose' / coin_type' / account' / change / address_index

If I'm right, the tests try to calculate a base index with a value greater than 0 instead of 0, and it should not be possible.

Are you ok to change the baseAddress constant value and index to 0 ?

Is it possible to introduce a currentAddress property to represent the address matching the index in the derivation path of the Account ?

Best regards,

Bruno

**Comment by satran004** [comment]
@bcoste Thanks a lot for the PR. It's a good enhancement.

>> This is due to my changes in the baseAddress, I force to calculate it from the external address of index 0.

I am wondering if we really need to force it to index 0. I noticed you are forcing it to 0 in two places,  during baseAddress generation and sign().  

If there is a utxo from a different address (let's say account=0, index=1) which is part of the input, I am not sure if the sign() will work. Because, you may need to sign the transaction with a private key from the address at index=1 to spend the fund. That's my understanding. 


>>From my understanding, base address represents the base index 0 of address_index in a BIP44 (and also CIP 1852) path :
m / purpose' / coin_type' / account' / change / address_index


As per my understanding, baseAddress where header_type is "0000" contains both paymentKey and StakeKey. But it doesn't talk about derivation path.  (plz check this cip: https://cips.cardano.org/cips/cip19/)

The baseAddress section in this doc also doesn't mention about cip1852 index = 0. https://docs.cardano.org/core-concepts/cardano-addresses

Same with cardano-serialization-lib rust impl https://github.com/Emurgo/cardano-serialization-lib/blob/6a8c67a0fdb2ac6d72132d9693b87a6734f3fed5/rust/src/address.rs#L488

>> Is it possible to introduce a currentAddress property to represent the address matching the index in the derivation path of the Account ?

I am not sure if it's good idea to introduce another address type "currentAddress". But if my understanding of baseAddress is not correct, we can definitely introduce that.


I just tried changing getExternalKeyPair() method to getHdKeyPair() in both Account.baseAddress(), Account.sign(). All tests are working fine after that change. I also tried a txn with utxos from index=2 and txn signing with HdPrivateKey at index=2.

Am I missing something?

File: src/main/java/com/bloxbean/cardano/client/account/Account.java
Comment (satran004) [inline src/main/java/com/bloxbean/cardano/client/account/Account.java:328]: Also,  we can change this method to private as it's not used  outside of this class. 
Currently if you print "Account" object, it shows privatekey and publickey in the log which may not be desirable.  If we really need to make this method public, we can add @JsonIgnore annotation to this method.

**Comment by satran004** [comment]
or, should we keep it simple for this PR. so for secret key, it is always account=0, index=0.

So, only one constructor in Account class for secret key. I don't know if we go with the current constructor with derivation path, if there will be any issue with changeAddress. May be, for this PR, we can keep it to index=0 for secret key and then extend it later as we learn more.

```
 public Account(Network network, String secretKey)
```

File: src/main/java/com/bloxbean/cardano/client/account/Account.java
Comment (bcoste) [reply src/main/java/com/bloxbean/cardano/client/account/Account.java:328]: I left this method public because I wanted to extract the public key to sign payloads of a transaction to give it to constructionCombine of cardano-rosetta.
But you're right, it's better to keep it private to have a good implementation.

**Comment by bcoste** [comment]
Thanks for your fast answer !
After checking the cardano docs and rust code, I agree with you on the fact a baseAddress can have an address-index index different from 0.
The changes you made are also fine from my point of view.
If we can just remove DerivationPath.createAccountDerivationPath() because I think it's not useful and may give wrong results ?

**Comment by satran004** [comment]
@bcoste Good to know that we are in sync now :)

It would be great if you can update the PR as discussed.  I will merge it after that.

- Remove derivation path from the new constructor for secret key in Account class
- Use getHdKey() in baseAddress() and sign() in Account class
- Remove getExternalHDKey() if not required.
- Remove any other methods which are not required. (Exp: DerivationPath.createAccountDerivationPath())


Thanks again for your contribution.

**Review by satran004 — APPROVED** [review]
Thanks @bcoste. Looks good to me. I will do some minor refactoring after merge. Like account secret key as byte[] in constructor instead of String.

**Pull request merged**
